includes/class-llms-order-generator.php

```diff
@@ -0,0 +1,597 @@
+	 */
+	public function get_gateway() {
+		return $this->gateway;
+	}
+
+	/**
+	 * Retrieves the order id to use for the order.
+	 *
+	 * Attempts to locate an existing pending order by order key if it was submitted,
+	 * otherwise returns `new` which denotes a new order should be created.
+	 *
+	 * @since [version]
+	 *
+	 * @return integer|string
+	 */
+	protected function get_order_id() {
+
+		$key      = $this->data['llms_order_key'] ?? null;
+		$order_id = null;
+
+		if ( $key ) {
+			$locate   = llms_get_order_by_key( $key, 'id' );
+			$order_id = $locate;
+		}
+
+		return $order_id ? $order_id : 'new';
+
+	}
+
+	/**
+	 * Retrieves the access plan for the order.
+	 *
+	 * @since [version]
+	 *
+	 * @return LLMS_Access_Plan|null
+	 */
+	public function get_plan() {
+		return $this->plan;
+	}
+
+	/**
+	 * Retrieves the student for the order.
+	 *
+	 * @since [version]
+	 *
+	 * @return LLMS_Student|null
+	 */
+	public function get_student() {
+		return $this->student;
+	}
+
+	/**
+	 * Retrieves an array of data representing the student.
+	 *
+	 * The resulting array is intended to be used for setting up the `LLMS_Order` post's
+	 * user metadata, ideally passed to `LLMS_Order::init()`.
+	 *
+	 * @since [version]
+	 *
+	 * @return array
+	 */
+	public function get_user_data() {
+
+		$map = array(
+			'billing_email'      => 'email_address',
+			'billing_first_name' => 'first_name',
+			'billing_last_name'  => 'last_name',
+			'billing_phone'      => 'llms_phone',
+		);
+
+		$data = array(
+			'billing_email'      => '',
+			'billing_first_name' => '',
+			'billing_last_name'  => '',
+			'billing_address_1'  => '',
+			'billing_address_2'  => '',
+			'billing_city'       => '',
+			'billing_state'      => '',
+			'billing_zip'        => '',
+			'billing_country'    => '',
+			'billing_phone'      => '',
+		);
+
+		foreach ( $data as $key => &$val ) {
+			$data_key = $map[ $key ] ?? "llms_{$key}";
+			$val = $this->data[ $data_key ] ?? '';
+		}
+
+		$data['user_id'] = $this->student ? $this->student->get( 'id' ) : '';
+
+		return $data;
+
+
+	}
+
+	/**
+	 * Performs all required data validations necessary to create the order.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean|WP_Errors Returns `true` if all validations pass or an error object.
+	 */
+	protected function validate() {
+
+		/**
+		 * Allows 3rd party validation prior to generation of an order.
+		 *
+		 * This validation hooks runs prior to all default validation.
+		 *
+		 * @since [version]
+		 *
+		 * @param null|WP_Error $validation_errors Halts checkout and returns the supplied error.
+		 */
+		$before_validation = apply_filters( 'llms_before_generate_order_validation', null );
+		if ( is_wp_error( $before_validation) ) {
+			return $before_validation;
+		}
+
+		$validations = array(
+			'validate_plan',
+			'validate_coupon',
+			'validate_gateway',
+			'validate_terms',
+			'validate_user',
+		);
+
+		foreach ( $validations as $func ) {
+			$res = $this->{$func}();
```

Comment:
oh my lord, someone finally stopped using `call_user_func(_array)`.

Reply:
There's still time to change this, tread carefully.